File: purser/protocol/src/Message.cpp
#include <stdio.h>

#include "Message.h"
#include "Log.h"
#include "MyException.h"

using namespace std;

Message::Message(const char* buf)
{
	Parse(buf);
}

void Message::Parse(const char* buf)
{
	char signature = buf[0];

	if(signature != PROTOCOL_SIGNATURE)
	{
		throw ExceptionProtocolError(PRINT_WHERE, "Wrong protocol signature");
	}

	mPhoneLen = buf[1];
	mTextLen = buf[2];
	char crc = buf[3];

	for(unsigned i = PHONE_OFFSET; i < PHONE_OFFSET + mPhoneLen; i++)
	{
	   mPhone += buf[i];
	}

	for(unsigned i = TEXT_OFFSET; i < TEXT_OFFSET + mTextLen; i++)
	{
	   mText += buf[i];
	}

	if(crc != GetCRC(mText))
	{
		throw ExceptionProtocolError(PRINT_WHERE, "Wrong CRC");
	}

	string str(buf);
}

string Message::Serialize() const
{
	string str;

	char crc = GetCRC(mText);

	str += PROTOCOL_SIGNATURE;
    str += char(mPhone.length());
	str += char(mText.length());
	str += crc;
	str += mPhone;
	str += string(MAX_PHONE_LEN - mPhone.length(), '\0');
	str += mText;
	str += string(MAX_TEXT_LEN - mText.length(), '\0');

	return str;
}


char Message::GetCRC(const string& str) const
{
    char crc = 0;
    for(char c: str)
    {
        crc ^= c;
    }

    return crc;
}

File: trunk/purser_java/Purser/src/com/tsi/purser/protocol/Message.java
package com.tsi.purser.protocol;

import com.tsi.data.Log;
import com.tsi.purser.exceptions.ProtocolException;



public class Message 
{	
	private int mPhoneLen;
	private int mTextLen;
	private String mPhone = new String("");
	private String mText = new String("");
	
	private final char ProtocolSignature = 'Z';
	private final int MaxPhoneLen = 32;
	private final int MaxTextLen = 220;
	private final int PhoneOffset = 4;
	private final int TextOffset = PhoneOffset + MaxPhoneLen;
	private final int MessageSize = PhoneOffset + MaxPhoneLen + MaxTextLen;

	
	public String GetPhone() { return mPhone; }
	public String GetText() { return mText; }
	
	public void SetPhone(String phone) { mPhone = phone; }
	public void SetText(String text) { mText = text; }
	
	
	public Message() {}
	
	public Message(String str) throws ProtocolException
	{
		Parse(str);
	}

	
	void Parse(String str) throws ProtocolException
	{
		char signature = str.charAt(0);

		if(signature != ProtocolSignature)
		{
			throw new ProtocolException("Wrong protocol signature");
		}

		mPhoneLen = str.charAt(1);
		mTextLen = str.charAt(2);
		char crc = str.charAt(3);
		
		Log.write("crc: " + crc);

		for(int i = PhoneOffset; i < PhoneOffset + mPhoneLen; i++)
		{
		   mPhone += str.charAt(i);
		}

		for(int i = TextOffset; i < TextOffset + mTextLen; i++)
		{
		   mText += str.charAt(i);
		}
		

		if(crc != GetCRC(mText))
		{
			throw new ProtocolException("Wrong CRC");
		}
	}
	
	
	
	public String Serialize() 
	{
		String str = "";

		char crc = GetCRC(mText);
		
		char[] zeroes = new char[MaxPhoneLen + MaxTextLen];
		for (int i=0; i< zeroes.length; i++) 
		{
			zeroes[i] = '0';
        }

		str += ProtocolSignature;
	    str += (char)mPhone.length();
		str += (char)mText.length();
		str += crc;
					
		str += mPhone;
		str += new String(zeroes, 0, MaxPhoneLen - mPhone.length());
		str += mText;
		str += new String(zeroes, 0, MaxTextLen - mText.length());
		
		Log.write("crc: " + crc);
		Log.write("mText: " + mText);

		return str;
	}
	

	private char GetCRC(String str) 
	{
	    char crc = 0;
	    for(int i=0; i<str.length(); i++)
	    {
	        crc ^= str.charAt(i);
	    }
	
	    return crc;
	}
}
